lib/google/gax/bundling.rb

```diff
@@ -0,0 +1,456 @@
+# Copyright 2016, Google Inc.
+# All rights reserved.
+#
+# Redistribution and use in source and binary forms, with or without
+# modification, are permitted provided that the following conditions are
+# met:
+#
+#     * Redistributions of source code must retain the above copyright
+# notice, this list of conditions and the following disclaimer.
+#     * Redistributions in binary form must reproduce the above
+# copyright notice, this list of conditions and the following disclaimer
+# in the documentation and/or other materials provided with the
+# distribution.
+#     * Neither the name of Google Inc. nor the names of its
+# contributors may be used to endorse or promote products derived from
+# this software without specific prior written permission.
+#
+# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+# "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+# LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+# A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+# OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+# SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+# LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+# DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+# THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+# (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+# OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+module Google
+  # Gax defines Google API extensions
+  module Gax
+    DEMUX_WARNING = [
+      'Warning: cannot demultiplex the bundled response, got ',
+      '%d subresponses; want %d, each bundled request will ',
+      'receive all responses'
+    ].join
+
+    # Helper function for #compute_bundle_id.
+    # Used to retrieve a nested field signified by name where dots in name
+    # indicate nested objects.
+    #
+    # @param obj [Object] an object.
+    # @param name [String] a name for a field in the object.
+    # @raise [NoMethodError] if the object does not contain the named attribute.
+    # @return [Object] value of named attribute. Can be nil.
+    def str_dotted_access(obj, name)
+      name.split('.').each do |part|
+        obj = obj.send(part)
```

Comment:
This will cause an exception if `obj` becomes `nil` in the middle.  `break if obj.nil?` might be necessary.

Also, Ruby protobuf accepts `[]`, so `obj = obj[part]` might be better (this would be helpful for testing, you can simply use a Hash for the data).